pkg/networkservice/common/connect/server.go

```diff
@@ -65,101 +84,205 @@ func WithDialOptionFactory(factory func(ctx context.Context, request *networkser
+	_, err = next.Server(ctx).Close(ctx, conn)
+
+	// Call client close
+	if _, closeErr := ce.client.Close(ctx, conn); closeErr != nil {
+		if err != nil {
+			// Combine errors
+			err = errors.Wrapf(err, "errors during close: %v", closeErr)
+		}
 	}
-	// TODO - do something with cancelFunc to clean up uRLToClientMap when we are done with them (maybe ref counting on Close?)
+
+	// remove connection from list
+	<-c.executor.AsyncExec(func() {
+		delete(ce.connections, conn.Id)
+		delete(c.connections, conn.Id)
+		// Close client if there is no more users for it.
+		c.closeClient(ctx, ce)
+	})
+
+	return &empty.Empty{}, err
+}
+
+func (c *connectServer) findOrCreateClient(ctx context.Context, conn *networkservice.Connection) (ce *clientEntry, err error) {
+	connectionExists := false
+	clientURI := clienturl.ClientURL(ctx)
+	<-c.executor.AsyncExec(func() {
+		if clientURI != nil {
+			ce, connectionExists = c.clientMap[clientURI.String()]
+		}
+		if !connectionExists {
+			// try to find using connection id
+			ce, connectionExists = c.connections[conn.Id]
+		}
+		if !connectionExists {
+			// Since there is no connection, we need to create it and mark as pending as well
+			// We could also be sure we one who will open connection.
+			if clientURI == nil {
+				ce.err = errors.Errorf("failed to create connection %v context should contain clienturl.ClientURL", conn)
+				return
+			}
+			ce = &clientEntry{
+				clientURI:   clientURI,
+				ready:       make(chan bool, 1),
+				connections: map[string]*networkservice.Connection{},
+			}
+			// Put/Update connection
+			c.clientMap[clientURI.String()] = ce
+		}
+		ce.connections[conn.Id] = conn
+	})
+	if ce.err != nil {
+		// Mark as ready but with error
+		ce.markAsReady(false)
+		c.closeClient(ctx, ce)
+		return nil, err
+	}
+	if connectionExists {
+		// Wait until connection is ready, or context timeout
+		select {
+		case <-ctx.Done():
+			// We failed, we need to remove pending state and close connection if we are last one.
+			<-c.executor.AsyncExec(func() {
+				delete(ce.connections, conn.Id)
+				c.closeClient(ctx, ce)
+			})
+			return nil, errors.Errorf("context timeout")
+		case <-ce.ready:
+			// all is fine, just return
+			if ce.err != nil {
+				<-c.executor.AsyncExec(func() {
+					delete(ce.connections, conn.Id)
+					c.closeClient(ctx, ce)
+				})
+				return nil, ce.err
+			}
+		}
+	} else {
+		// Dial and create client connection
+		if err := c.createClient(ctx, ce); err != nil {
+			// If we failed to create client, we failed to dial, so no need to close,
+			// we need to mark it as error one and all clients pending will return errors, and last one will remove entry
+			// Mark connection as ready to use
+			<-c.executor.AsyncExec(func() {
+				ce.err = err
+				// Mark as ready but with error
+				ce.markAsReady(false)
+				c.closeClient(ctx, ce)
+			})
+			return nil, err
+		}
+
+		// Mark connection as ready to use
+		<-c.executor.AsyncExec(func() {
```

Comment:
I don't quite follow why we are wrapping this in an AsyncExec for the whole server. It looks like the ce objects are marked as ready by their creator, and thus we don't need to get exclusivity at the server level for this operation.

Reply:
Fixed.